payment_gateway_adyen/services/payment_service.py

```diff
@@ -0,0 +1,244 @@
+            ('payment_mode_id.provider', '=', 'adyen'),
+            ])
+        if transaction:
+            transaction.write(vals)
+        else:
+            raise UserError(
+                _('The transaction %s do not exist in Odoo') % result.message[
+                    'pspReference'])
+
+    def _get_formatted_amount(self, amount=None):
+        if amount is None:
+            amount = self.collection._get_amount_to_capture()
+        if self.collection.currency_id.name in ZERO_DECIMAL_CURRENCIES:
+            return int(amount)
+        elif self.collection.currency_id.name in THREE_DECIMAL_CURRENCIES:
+            return int(float_round(amount * 1000, 0))
+        else:
+            return int(float_round(amount * 100, 0))
+
+    def _prepare_charge(self, source=None, **kwargs):
+        transaction = self.collection
+        description = "|".join([
+            transaction.name,
+            transaction.partner_id.email,
+            str(transaction.id)])
+        vals = {
+            'amount': {
+                "value": self._get_formatted_amount(),
+                "currency": transaction.currency_id.name},
+            'reference': description,
+            'additionalData': {"executeThreeD": "true"}  # optional may be?
+            }
+        if kwargs.get('encrypted_card'):
+            vals['additionalData'] = {
+                'card.encrypted.json': kwargs['encrypted_card'],
+                'executeThreeD': 'true'
+                }
+#        elif kwargs.get('card'):
+#            # PCI clear card data
+#            vals['card'] = kwargs['card']
+        else:
+            raise UserError(_('Error: card information missing!'))
+        return vals
+
+    def _get_adyen_client(self):
+        account = self._get_account()
+        data = account.get_data()
+        ady = Adyen.Adyen()
+        ady.payment.client.username = data['username']
+        ady.payment.client.platform = str(data['platform'])
+        ady.payment.client.merchant_account = data['merchant_account']
+        ady.payment.client.password = account._get_password()
+        ady.payment.client.app_name = data['app_name']
+        return ady.payment
+
+    def _create_transaction(self, source=None, **kwargs):
+        payload = self._prepare_charge(source=source, **kwargs)
+        return self._http_adyen_request('authorise', payload)
+
+    def _parse_creation_result(self, transaction, **kwargs):
+        # TODO does it make sense to call super with the specific AdyenResult
+        # object?
+        res = super(PaymentService, self).\
+            _parse_creation_result(transaction, **kwargs)
+        transaction = transaction.message
+        res.update({
+            'amount': self.collection._get_amount_to_capture(),
+            'external_id': transaction['pspReference'],
+            'state': MAP_SOURCE_STATE[transaction['resultCode']],
+            'data': json.dumps(transaction),
+            })
+        # TODO risk analysis if possible with API
```

Comment:
creation do not return information on risk so you can remove it

Reply:
ok